File: src/day1/BTBFS.ts
//    Breadth First Search / Traversal
//    using implicit Data Structure -> Queue
//           7             <- root / head
//         /   \               |
//       23      3             |
//     /   \   /   \           |
//    5    4   18   21     <- leaf
// 
// in_order_search => [7, 23, 3, 5, 4, 18, 21]
export default function bfs(head: BinaryNode<number>, needle: number): boolean {

    const queue: (BinaryNode<number> | null)[] = [head]
    
    while (queue.length) {

        const current = queue.shift() as BinaryNode<number> | undefined | null;
        if (!current) {
            continue
        }

        // search
        if (current.value === needle) {
            return true
        }

        queue.push(current.left)
        queue.push(current.right)
    }

    return false;
}

File: src/day1/BTBFS.go
package main

//	Breadth First Search / Traversal
//	using implicit Data Structure -> Queue
//	       7             <- root / head
//	     /   \               |
//	   23      3             |
//	 /   \   /   \           |
//	5    4   18   21     <- leaf
//
// in_order_search => [7, 23, 3, 5, 4, 18, 21]
func BFS(head *BinaryNode, needle int) bool {
	queue := []*BinaryNode{head}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if current == nil {
			continue
		}

		// search
		if current.Value == needle {
			return true
		}

		queue = append(queue, current.Left)
		queue = append(queue, current.Right)
	}

	return false
}
